packages/zoe/src/zoeService/purseStorage.js

```diff
@@ -0,0 +1,103 @@
+// @ts-check
+
+import { amountMath } from '@agoric/ertp';
+import { E } from '@agoric/eventual-send';
+import { makeWeakStore as makeNonVOWeakStore } from '@agoric/store';
+
+import { arrayToObj, objectMap } from '../objArrayConversion';
+
+export const makePurseStorage = () => {
+  /** @type {WeakStore<Brand, ERef<Purse>>} */
+  const brandToPurse = makeNonVOWeakStore('brand');
+
+  /**
+   * Create a purse for a new issuer
+   *
+   * @param {IssuerRecord} record
+   * @returns {void}
+   */
+  const createPurse = record => {
+    if (!brandToPurse.has(record.brand)) {
+      brandToPurse.init(record.brand, E(record.issuer).makeEmptyPurse());
+    }
+  };
+
+  /**
+   * Create a purse for a new, local issuer. Used only for ZCFMint issuers.
+   *
+   * @param {IssuerRecord} record
+   * @returns {Purse}
+   */
+  const createLocalPurse = record => {
+    const localPurse = record.issuer.makeEmptyPurse();
+    brandToPurse.init(record.brand, localPurse);
+    return localPurse;
+  };
+
+  /** @type {WithdrawPayments} */
+  const withdrawPayments = allocation => {
+    return harden(
+      objectMap(allocation, ([keyword, amount]) => {
+        const purse = brandToPurse.get(amount.brand);
+        return [keyword, E(purse).withdraw(amount)];
+      }),
+    );
+  };
+
+  /**
+   *
+   *  Only used internally. Actually deposit a payment or promise for payment.
+   *
+   * @param {ERef<Payment>} paymentP
+   * @param {Amount} amount
+   * @returns {Promise<Amount>}
+   */
+  const doDepositPayments = (paymentP, amount) => {
+    const purse = brandToPurse.get(amount.brand);
+    return E.when(paymentP, payment => E(purse).deposit(payment, amount));
+  };
+
+  // Proposal is cleaned, but payments are not
+
+  /**
+   * Deposits payments or promises for payments according to the
+   * `give` property of the proposal. Using the proposal, creates an
+   * initial allocation including the amount deposited for `give`
+   * keywords and an empty amount for `want` keywords.
+   *
+   * @param {ProposalRecord} proposal
+   * @param {PaymentPKeywordRecord} payments
+   * @returns {Promise<Allocation>}
+   */
+  const depositPayments = async (proposal, payments) => {
+    const { give, want } = proposal;
+    const giveKeywords = Object.keys(give);
+    const wantKeywords = Object.keys(want);
+    const proposalKeywords = harden([...giveKeywords, ...wantKeywords]);
+
+    const amountsDeposited = await Promise.all(
+      giveKeywords.map(keyword =>
+        doDepositPayments(payments[keyword], give[keyword]),
+      ),
+    );
+
+    const amountsWanted = wantKeywords.map(keyword =>
+      amountMath.makeEmptyFromAmount(want[keyword]),
+    );
+
+    // Note: payments without a matching `give` keyword are ignored.
```

Comment:
Thinking about the possibly optional `give`, could it make sense that payments without matching `give` keywords add new `give` entries? Among the advantages is that these funds would be recovered from the payouts, where the user's code is more likely to look for them than looking for unconsumed elements of the offer.

Reply:
We could do that, but a payment without a matching `give` would be a mistake currently, and it seems bad to silently handle the mistake and rely on the payout mechanism for the refund. I think I would like to error immediately and give a refund of all payments if that happens.

Reply:
An error would be great! Much better than silently ignoring them. Can we have an error?

Reply:
Btw, thanks for the note comment. Without it, I don't know that I would have spotted the hazard.

Reply:
> An error would be great! Much better than silently ignoring them. Can we have an error?

I think so. I think we would need to clean the paymentKeywordRecord keywords and compare them to `give` and throw if they are not exactly the same. Giving refunds is harder, but maybe we could check for any additional payment keywords first before depositing, and then just return the payment keyword record. Or maybe we need to return a user seat with the returned paymentKeywordRecord as the payout. This sounds a bit complicated though, so I'd like to make an issue and do this in a separate PR, if that's ok.

Reply:
Added an issue here: https://github.com/Agoric/agoric-sdk/issues/2889

Reply:
Added issue is fine. Thanks. Resolving.